internal/mempool/v1/mempool.go

```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
-			txmp.removeTx(wtx, false)
-		}
+		// Regardless of success, remove the transaction from the mempool.
+		_ = txmp.removeTxByKey(tx.Key())
 	}
 
 	txmp.purgeExpiredTxs(blockHeight)
 
 	// If there any uncommitted transactions left in the mempool, we either
 	// initiate re-CheckTx per remaining transaction or notify that remaining
 	// transactions are left.
-	if txmp.Size() > 0 {
+	size := txmp.Size()
+	txmp.metrics.Size.Set(float64(size))
+	if size > 0 {
 		if txmp.config.Recheck {
-			txmp.logger.Debug(
-				"executing re-CheckTx for all remaining transactions",
-				"num_txs", txmp.Size(),
-				"height", blockHeight,
-			)
-			txmp.updateReCheckTxs()
+			txmp.recheckTransactions()
 		} else {
 			txmp.notifyTxsAvailable()
 		}
 	}
-
-	txmp.metrics.Size.Set(float64(txmp.Size()))
 	return nil
 }
 
-// initTxCallback performs the initial, i.e. the first, callback after CheckTx
-// has been executed by the ABCI application. In other words, initTxCallback is
-// called after executing CheckTx when we see a unique transaction for the first
-// time. CheckTx can be called again for the same transaction at a later point
-// in time when re-checking, however, this callback will not be called.
-//
-// After the ABCI application executes CheckTx, initTxCallback is called with
-// the ABCI *Response object and TxInfo. If postCheck is defined on the mempool,
-// we execute that first. If there is no error from postCheck (if defined) and
-// the ABCI CheckTx response code is OK, we attempt to insert the transaction.
+// initTxCallback handles the ABCI CheckTx response for the first time a
+// transaction is added to the mempool.  A recheck after a block is committed
+// goes to the default callback (see recheckTxCallback).
 //
-// When attempting to insert the transaction, we first check if there is
-// sufficient capacity. If there is sufficient capacity, the transaction is
-// inserted into the txStore and indexed across all indexes. Otherwise, if the
-// mempool is full, we attempt to find a lower priority transaction to evict in
-// place of the new incoming transaction. If no such transaction exists, the
-// new incoming transaction is rejected.
+// If either the application rejected the transaction or a post-check hook is
+// defined and rejects the transaction, it is discarded.
 //
-// If the new incoming transaction fails CheckTx or postCheck fails, we reject
-// the new incoming transaction.
+// Otherwise, if the mempool is full, check for lower-priority transactions
+// that can be evicted to make room for the new one. If no such transactions
+// exist, this transaction is logged and dropped; otherwise the selected
+// transactions are evicted.
 //
-// NOTE:
-// - An explicit lock is NOT required.
-func (txmp *TxMempool) initTxCallback(wtx *WrappedTx, res *abci.Response, txInfo mempool.TxInfo) {
+// Finally, the new transaction is added and size stats updated.
+func (txmp *TxMempool) initTxCallback(wtx *WrappedTx, res *abci.Response) {
```

Comment:
```suggestion
func (txmp *TxMempool) initialTxCallback(wtx *WrappedTx, res *abci.Response) {
```
nit: when I see init I'm too used to thinking initialize